### ISSUE TITLE
Unify supervised & unsupervised flow

### ISSUE THREAD

**Issue by nviwch** [LiheYoung/UniMatch-V2#10]
Great work. I found that in the paper you form a batch by keeping equal number of labeled & unlabeled images. Using OHEM for labeled images and UniMatch v2 for unlabeled images.

I wonder if you have tested unified flow which is to use UniMatch v2 flow no matter labeled or unlabeled images, keep using pseudo-label for unlabeled images while using true label for labeled images (which means labeled images will not involve f_ema).

I am thinking if the unified flow stated above would make the performance better or worse. I think the concept of comparing perturbation may be beneficial to labeled images too and unified flow is even simpler and more elegant.

Moreover, the unsupervised loss is using pseudo-label which is argmax of weakly augmented prediction, have you tried to use the prediction distribution directly just like knowledge distillation?

**Comment by LiheYoung** [comment]
This is a good question. We have one prior work showing that applying strong perturbations to labeled images are generally destructive to the performance. You may refer to the Table 5 in [ST++](https://arxiv.org/pdf/2106.05095) for details. We have also tried to apply strong perturbations to labeled images under the UniMatch framework, but it does not bring better performance. This practice may help in some extremely label-scarce regimes (*e.g.,*, only 50 labeled images), but in general, it can not benefit the training.

A intuitive explanation for this is that the human annotations are already a form of perturbation for the network. Human annotations are more challenging to learn than pseudo labels produced by the model itself. Therefore, additional hand-crafted input-level or feature-level perturbations are not necessary. Another point for this is that, we need to provide some clean training images (*i.e.,* without perturbations) for the model to learn as well, to mitigate the distribution shift between strongly augmented training images and clean test images.

**Comment by nviwch** [comment]
Thank you for your reply. 

Just sharing the result using my own dataset (which is very different from public dataset), I am using three cross-entropy losses in total.
1. Weak augmentation + ground truth
2. Strong augmentation + ground truth
3.  Strong augmentation + pseudo label (online generated with weak augmentation)

For loss number 2, having strong augmentation with labeled data can boost the performance. 
For loss number 3, using weak augmentation prediction with thresholding (as am doing multi-label segmentation, I used > 0.5 to create pseudo-label instead of argmax stated in the paper) will hinder the performance significantly. To dig deeper, I tried to use the distribution directly with detach and the result is equally bad. In other words, I need to keep the gradient flow through both strong & weak augmentation unsupervised branch for good performance.
